File: app/src/main/java/com/example/android/popularmovies/adapter/MoviesAdapter.java
package com.example.android.popularmovies.adapter;

import android.content.Context;
import androidx.databinding.DataBindingUtil;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.example.android.popularmovies.R;
import com.example.android.popularmovies.activity.MovieDetails;
import com.example.android.popularmovies.databinding.ListItemsBinding;
import com.example.android.popularmovies.model.Movie;

import java.util.List;


public class MoviesAdapter extends RecyclerView.Adapter<MoviesAdapter.MovieViewHolder> {


    private List<Movie> movies;
    private Context context;

    class MovieViewHolder extends RecyclerView.ViewHolder {

        private ListItemsBinding itemsBinding;

        MovieViewHolder(ListItemsBinding listItemsBinding) {
            super(listItemsBinding.getRoot());
            this.itemsBinding = listItemsBinding;
        }
    }

    public MoviesAdapter(List<Movie> movies) {
        this.movies = movies;

    }

    @NonNull
    @Override
    public MoviesAdapter.MovieViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        context = parent.getContext();
        LayoutInflater layoutInflater = LayoutInflater.from(context);
        ListItemsBinding listItemsBinding = DataBindingUtil.inflate(layoutInflater, R.layout.list_items, parent, false);

        return new MovieViewHolder(listItemsBinding);
    }

    @Override
    public void onBindViewHolder(@NonNull final MovieViewHolder holder, final int position) {

        holder.itemsBinding.setMovie(movies.get(position));
        holder.itemsBinding.setClickHandler(new ClickHandlers());
    }

    @Override
    public int getItemCount() {
        return movies.size();
    }


    public class ClickHandlers {
        public void onClickMovie(Movie movie) {
            MovieDetails.start(context, movie);
        }
    }

    public void addItem(List<Movie> movies) {
        this.movies.addAll(movies);
        notifyDataSetChanged();
    }
}

File: app/src/main/kotlin/com/example/android/popularmovies/adapter/MoviesAdapter.kt
package com.example.android.popularmovies.adapter

import android.content.Context
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup

import com.example.android.popularmovies.R
import com.example.android.popularmovies.activity.MovieDetailsActivity
import com.example.android.popularmovies.databinding.ListItemsBinding
import com.example.android.popularmovies.model.Movie


class MoviesAdapter(private val movies: MutableList<Movie>) : RecyclerView.Adapter<MoviesAdapter.MovieViewHolder>() {
    private var context: Context? = null

    inner class MovieViewHolder(val itemsBinding: ListItemsBinding) : RecyclerView.ViewHolder((itemsBinding as ViewDataBinding).root)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieViewHolder {
        context = parent.context
        val layoutInflater = LayoutInflater.from(context)
        val listItemsBinding = DataBindingUtil.inflate<ViewDataBinding>(layoutInflater, R.layout.list_items, parent, false) as ListItemsBinding

        return MovieViewHolder(listItemsBinding)
    }

    override fun onBindViewHolder(holder: MovieViewHolder, position: Int) {

        holder.itemsBinding.movie = movies[position]
        holder.itemsBinding.clickHandler = ClickHandlers()
    }

    override fun getItemCount(): Int {
        return movies.size
    }


    inner class ClickHandlers {
        fun onClickMovie(movie: Movie) {
            context?.let { MovieDetailsActivity.start(it, movie) }
        }
    }

    fun addItem(movies: List<Movie>) {
        this.movies.addAll(movies)
        notifyDataSetChanged()
    }
}

